File: algorithms/2_medium/R/00402_remove_k_digits/stack.c
typedef ptrdiff_t isize;

static inline isize min_isize(const isize a, const isize b)
{
    return a < b ? a : b;
}

char* removeKdigits(char* num, int k)
{
    assert(num != NULL && num[0] != '\0');
    isize stack_len = 0;

    for (isize i = 0; num[i] != '\0'; i += 1) {
        while (k > 0 && stack_len > 0 && num[stack_len - 1] > num[i]) {
            stack_len -= 1;
            k -= 1;
        }

        if (stack_len == 0 && num[i] == '0') {
            continue;
        }

        num[stack_len] = num[i];
        stack_len += 1;
    }

    stack_len -= min_isize(stack_len, k);

    if (stack_len == 0) {
        num[stack_len] = '0';
        stack_len += 1;
    }

    num[stack_len] = '\0';

    return num;
}

File: algorithms/2_medium/R/00402_remove_k_digits/stack.rs
impl Solution {
    pub fn remove_kdigits(num: String, mut k: i32) -> String {
        let mut bytes = num.into_bytes();
        assert!(bytes.len() > 0);
        let mut stack_len = 0_usize;

        for i in 0..bytes.len() {
            while k > 0 && stack_len > 0 && bytes[stack_len - 1] > bytes[i] {
                stack_len -= 1;
                k -= 1;
            }

            if stack_len == 0 && bytes[i] == b'0' {
                continue;
            }

            bytes[stack_len] = bytes[i];
            stack_len += 1;
        }

        stack_len -= std::cmp::min(stack_len, k as usize);

        if stack_len == 0 {
            bytes[stack_len] = b'0';
            stack_len += 1;
        }

        bytes.truncate(stack_len);

        String::from_utf8(bytes).unwrap()
    }
}
